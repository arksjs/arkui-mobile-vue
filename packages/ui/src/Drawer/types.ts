import type { PlacementType } from '../helpers'
import type { PopupEmits, PopupProps } from '../popup/types'

export interface DrawerProps extends PopupProps {
  title?: string
  placement?: PlacementType
  showClose?: boolean
  showMask?: boolean
  enableSafeAreaInsets?: boolean
}

export type DrawerEmits = PopupEmits
