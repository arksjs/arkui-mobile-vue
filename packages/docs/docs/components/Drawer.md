# Drawer 抽屉

用于在屏幕边缘显示应用导航等内容的面板。

<CodeDemo name="Drawer">

<<< @/../../demo/src/components/Basic/Drawer/index.vue

</CodeDemo>

## Import

```js
import { Drawer } from 'vfox'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

### Import Type

组件导出的类型定义：

```ts
import type {
  VisibleState,
  PopupOnVisibleStateChange,
  PopupOnCancel
} from 'vfox'
```

## Props

| 属性            | 类型          | 默认值   | 必填 | 说明                                               |
| --------------- | ------------- | -------- | ---- | -------------------------------------------------- |
| v-model:visible | boolean       | false    | 否   | 是否显示                                           |
| title           | string        |          | 否   | 标题，设置标题后展示头部栏                         |
| placement       | PlacementType | 'bottom' | 否   | 从哪展开，可选值：'bottom', 'top', 'left', 'right' |
| show-close      | boolean       | false    | 否   | 是否显示关闭按钮，显示按钮后展示头部栏             |

## Events

| 事件                 | 描述                           | 回调函数参数                                                        | TypeScript 函数           |
| -------------------- | ------------------------------ | ------------------------------------------------------------------- | ------------------------- |
| visible-state-change | 展示隐藏时触发                 | payload: { state: [VisibleState](./Drawer.md#visiblestate-值说明) } | PopupOnVisibleStateChange |
| cancel               | 关闭按钮或者点击蒙层关闭时触发 | payload: { source: string }                                         | PopupOnCancel             |

### VisibleState 值说明

| 值     | 说明                 | 备注                                              |
| ------ | -------------------- | ------------------------------------------------- |
| show   | 展示时触发           |                                                   |
| shown  | 展示且动画结束后触发 |                                                   |
| hide   | 隐藏时触发           | 可能携带其他参数 cancel, maskClick, closeClick 等 |
| hidden | 隐藏且动画结束后触发 | 可能携带其他参数 cancel, maskClick, closeClick 等 |

## Slots

```vue
<fx-drawer title="菜单" placement="right">
  内容
</fx-drawer>
```
