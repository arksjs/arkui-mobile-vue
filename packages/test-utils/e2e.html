<script src="../../node_modules/vue/dist/vue.global.js"></script>
<script src="../../dist/index.js"></script>
<link rel="stylesheet" href="../../dist/index.css" />

<div id="app"></div>
