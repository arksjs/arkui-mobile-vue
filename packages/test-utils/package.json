{
  "name": "@arksjs/test-utils",
  "version": "0.0.1",
  "description": "arkui test-utils",
  "main": "dist/index.js",
  "private": true
}
