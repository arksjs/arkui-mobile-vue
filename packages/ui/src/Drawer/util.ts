import { PLACEMENT_TYPES, getEnumsValue, type CSSProperties, type PlacementType } from '../helpers'
import type { SafeAreaInsets } from '../hooks'

export const getInnerStyles = ({
  placement,
  safeAreaInsets
}: {
  placement?: PlacementType
  safeAreaInsets: SafeAreaInsets
}) => {
  const _placement = getEnumsValue(PLACEMENT_TYPES, placement)

  let left = safeAreaInsets.left
  let top = safeAreaInsets.top
  let right = safeAreaInsets.right
  let bottom = safeAreaInsets.bottom

  if (_placement === 'top') {
    bottom = 0
  } else if (_placement === 'bottom') {
    top = 0
  } else if (_placement === 'left') {
    right = 0
  } else if (_placement === 'right') {
    left = 0
  }

  return {
    padding: top + 'px ' + right + 'px ' + bottom + 'px ' + left + 'px'
  } as CSSProperties
}

export const getInnerClasses = ({
  placement,
  hasHeader
}: {
  placement?: PlacementType
  hasHeader: boolean
}) => [
  'ta-drawer_inner',
  'ta-horizontal-hairline',
  'placement--' + getEnumsValue(PLACEMENT_TYPES, placement),
  {
    'has--header': hasHeader
  }
]
