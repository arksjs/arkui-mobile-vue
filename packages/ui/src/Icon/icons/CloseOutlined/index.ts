export { default as default } from './CloseOutlined.vue'
