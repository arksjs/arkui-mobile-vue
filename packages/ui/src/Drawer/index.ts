import { withInstall } from '../helpers/with-install'
import Drawer from './Drawer.vue'

export { Drawer }
export const install = withInstall(Drawer)
export default Drawer
