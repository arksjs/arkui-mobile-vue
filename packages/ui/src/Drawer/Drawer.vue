<template>
  <teleport to="body">
    <div :class="[popupClasses, classes]" :style="popupStyles" v-bind="$attrs">
      <div class="ta-mask" @click="onMaskClick"></div>
      <div :class="innerClasses" :style="innerStyles">
        <slot name="header">
          <NavBar
            v-if="hasHeader"
            class="ta-drawer_header"
            :title="title"
            :rightButtons="
              showClose ? [{ icon: CloseOutlined, text: 'close' }] : []
            "
            :iconOnly="true"
            @rightButtonClick="onHeaderRightClick"
          >
          </NavBar>
        </slot>
        <div class="ta-drawer_body">
          <slot></slot>
        </div>
      </div>
    </div>
  </teleport>
</template>

<script lang="ts">
import { defineComponent, computed, toRef, watch } from 'vue'
import type { PropType } from 'vue'
import { NavBar } from '../NavBar'
import { usePopup } from '../popup/use-popup'
import { popupEmits, popupProps } from '../popup/popup'
import { useSafeAreaInsets } from '../hooks/use-safe-area-insets'
import { createEnumsValidator, getEnumsValue } from '../helpers/validator'
import { PLACEMENT_TYPES } from '../helpers/constants'
import type { PlacementType, PropsToEmits } from '../helpers/types'
import CloseOutlined from '../Icon/icons/CloseOutlined'
import { getClasses, getInnerClasses, getInnerStyles } from './util'
import type { DrawerEmits } from './types'

export default defineComponent({
  name: 'ta-drawer',
  components: { NavBar },
  props: {
    ...popupProps,
    title: {
      type: String,
      default: null
    },
    placement: {
      type: String as PropType<PlacementType>,
      validator: createEnumsValidator(PLACEMENT_TYPES),
      default: getEnumsValue(PLACEMENT_TYPES)
    },
    showClose: {
      type: Boolean,
      default: false
    },
    // 是否开启安全区
    enableSafeAreaInsets: {
      type: Boolean,
      default: true
    },
    showMask: {
      type: Boolean,
      default: true
    }
  },
  emits: { ...popupEmits } as PropsToEmits<DrawerEmits>,
  setup(props, ctx) {
    const popup = usePopup(props, ctx, {})
    const { safeAreaInsets } = useSafeAreaInsets(
      toRef(props, 'enableSafeAreaInsets')
    )

    const hasHeader = computed(
      () => !!(props.title || props.showClose || ctx.slots.header)
    )

    const innerStyles = computed(() =>
      getInnerStyles({
        placement: props.placement,
        safeAreaInsets
      })
    )

    function onHeaderRightClick() {
      if (props.showClose) {
        popup.onCloseClick()
      }
    }

    watch(
      () => props.showMask,
      val => popup.setEnableBlurCancel(!val),
      { immediate: true }
    )

    const classes = computed(() => getClasses(props.showMask))
    const innerClasses = computed(() =>
      getInnerClasses({
        placement: props.placement,
        hasHeader: hasHeader.value
      })
    )

    return {
      ...popup,
      classes,
      innerClasses,
      hasHeader,
      innerStyles,
      onHeaderRightClick,
      CloseOutlined
    }
  }
})
</script>
