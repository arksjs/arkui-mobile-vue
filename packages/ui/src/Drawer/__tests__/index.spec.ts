import { describe, test, expect } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import { Drawer } from '@/Drawer'
import CloseOutlined from '@/Icon/icons/CloseOutlined'

describe('Drawer', () => {
  test('snapshot', () => {
    const wrapper = shallowMount(Drawer, {
      global: {
        stubs: { teleport: true }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('title', async () => {
    const title = 'title'
    const wrapper = mount(Drawer, {
      global: {
        stubs: { teleport: true }
      },
      props: {
        title
      }
    })
    expect(wrapper.find('.fx-nav-bar_title').text()).toBe(title)
    expect(wrapper.find('.fx-drawer_header').exists()).toBeTruthy()
  })

  test('showClose', async () => {
    const wrapper = mount(Drawer, {
      global: {
        stubs: { teleport: true }
      }
    })
    expect(wrapper.findComponent(CloseOutlined).exists()).toBeFalsy()
    expect(wrapper.find('.fx-drawer_header').exists()).toBeFalsy()

    await wrapper.setProps({
      showClose: true
    })
    expect(wrapper.findComponent(CloseOutlined).exists()).toBeTruthy()
    expect(wrapper.find('.fx-drawer_header').exists()).toBeTruthy()
  })

  test('placement', async () => {
    const wrapper = mount(Drawer, {
      global: {
        stubs: { teleport: true }
      }
    })
    expect(wrapper.find('.fx-drawer_inner').classes()).toContain(
      'placement--bottom'
    )

    await wrapper.setProps({
      placement: 'top'
    })
    expect(wrapper.find('.fx-drawer_inner').classes()).toContain(
      'placement--top'
    )

    await wrapper.setProps({
      placement: 'left'
    })
    expect(wrapper.find('.fx-drawer_inner').classes()).toContain(
      'placement--left'
    )

    await wrapper.setProps({
      placement: 'right'
    })
    expect(wrapper.find('.fx-drawer_inner').classes()).toContain(
      'placement--right'
    )
  })
})
