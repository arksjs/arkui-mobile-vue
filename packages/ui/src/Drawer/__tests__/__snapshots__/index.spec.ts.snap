// Vitest Snapshot v1

exports[`Drawer > snapshot 1`] = `
"<teleport-stub>
  <div class=\\"fx-popup fx-drawer\\" style=\\"z-index: 2000; display: none;\\">
    <div class=\\"fx-mask\\"></div>
    <div class=\\"fx-drawer_inner fx-horizontal-hairline placement--bottom\\" style=\\"padding: 0px 0px 0px 0px;\\">
      <!--v-if-->
      <div class=\\"fx-drawer_body\\"></div>
    </div>
  </div>
</teleport-stub>"
`;
